planner/core/rule_inject_extra_projection.go

```diff
@@ -125,3 +131,70 @@ func injectProjBelowAgg(aggPlan PhysicalPlan, aggFuncs []*aggregation.AggFuncDes
 	aggPlan.SetChildren(proj)
 	return aggPlan
 }
+
+// The schema of PhysicalSort and PhysicalTopN are the same as the schema of
+// their children. When a projection is injected as the child of
+// PhysicalSort and PhysicalTopN, some extra columns will be added into the
+// schema of the Projection, thus we need to add another Projection upon
+// them to prune the redundant columns.
+func injectProjBelowSort(p PhysicalPlan, orderByItems []*ByItems) PhysicalPlan {
```

Comment:
can we do this when we creating the `LogicalSort` operator in plan builder?

Reply:
Seems not, `projection` will forbid the push-down of TopN.